File: Stack And Queue/20.有效的括号.cpp
#include <string>

using namespace std;
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> result;

        for (char w : s)
        {
            // 将{ [ (的反括号存入栈
            if (w == '[')
                result.push(']');
            else if (w == '{')
                result.push('}');
            else if (w == '(')
                result.push(')');
            // 输入}])时栈为空，返回false
            else if (result.empty())
                return false;
            // 栈中返回栈顶元素, 如果相等进行去除，不相等直接判断false
            else if (w != result.top())
                return false;
            else
                result.pop();
        }

        return result.empty();
    }
};
// @lc code=end

File: Stack And Queue/20.有效的括号.ts
/*
 * @lc app=leetcode.cn id=20 lang=typescript
 *
 * [20] 有效的括号
 */

// @lc code=start
function isValid(s: string): boolean {
  // 此题适合用栈来处理，使用js中的数组模拟栈的行为
  let stack = []
  let arr = s.split("")

  for (let i = 0; i < arr.length; i++) {
    const length = stack.length
    const last = length ? stack[length - 1] : null

    switch (arr[i]) {
      case "(":
        stack.push(arr[i])
        break
      case "[":
        stack.push(arr[i])
        break
      case "{":
        stack.push(arr[i])
        break
      case ")":
        if (last !== "(") {
          return false
        }
        stack.pop()
        break
      case "]":
        if (last !== "[") {
          return false
        }
        stack.pop()
        break
      case "}":
        if (last !== "{") {
          return false
        }
        stack.pop()
        break
      default: break
    }
  }

  return stack.length === 0
};
// @lc code=end

